server/src/main/java/org/elasticsearch/index/mapper/SourceValueFetcher.java

```diff
@@ -0,0 +1,94 @@
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.search.lookup.SourceLookup;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * An implementation of {@link ValueFetcher} that knows how to extract values
+ * from the document source. Most standard field mappers will use this class
+ * to implement value fetching.
+ */
+public abstract class SourceValueFetcher implements ValueFetcher {
+    private final Set<String> sourcePaths;
+    private final @Nullable Object nullValue;
+    private final boolean parsesArrayValue;
+
+    public SourceValueFetcher(String fieldName, MapperService mapperService, boolean parsesArrayValue) {
+        this(fieldName, mapperService, parsesArrayValue, null);
+    }
+
+    /**
+     * @param fieldName The name of the field.
+     * @param parsesArrayValue Whether the fetcher handles array values during document parsing.
+     * @param nullValue A optional substitute value if the _source value is 'null'.
+     */
+    public SourceValueFetcher(String fieldName, MapperService mapperService, boolean parsesArrayValue, Object nullValue) {
+        this.sourcePaths = mapperService.sourcePath(fieldName);
+        this.nullValue = nullValue;
+        this.parsesArrayValue = parsesArrayValue;
+    }
+
+    @Override
+    public List<Object> fetchValues(SourceLookup lookup) {
+        List<Object> values = new ArrayList<>();
+        for (String path : sourcePaths) {
+            Object sourceValue = lookup.extractValue(path, nullValue);
+            if (sourceValue == null) {
+                return List.of();
+            }
+
+            if (parsesArrayValue) {
+                values.addAll((List<?>) parseSourceValue(sourceValue));
+            } else {
+                // We allow source values to contain multiple levels of arrays, such as `"field": [[1, 2]]`.
+                // So we need to unwrap these arrays before passing them on to be parsed.
+                Queue<Object> queue = new ArrayDeque<>();
+                queue.add(sourceValue);
+                while (queue.isEmpty() == false) {
+                    Object value = queue.poll();
+                    if (value instanceof List) {
+                        queue.addAll((List<?>) value);
+                    } else {
+                        Object parsedValue = parseSourceValue(value);
+                        if (parsedValue != null) {
+                            values.add(parsedValue);
+                        }
+                    }
+                }
+            }
+        }
+        return values;
+    }
+
+    /**
+     * Given a value that has been extracted from a document's source, parse it into a standard
+     * format. This parsing logic should closely mirror the value parsing in
+     * {@link FieldMapper#parseCreateField} or {@link FieldMapper#parse}.
+     */
+    protected abstract Object parseSourceValue(Object value);
```

Comment:
I don't really think this is cleaner than passing the `CheckedFunction` but if you like it better that is cool with me.

Reply:
I agree, I think passing a function argument is readable too. It seems totally fine if the implementation evolved in that direction.

I had a slight preference for this approach since it associates a name with the function argument: when you look at an implementation like `KeywordFieldMapper#valueFetcher`, the purpose of the function is very obvious.